8288415: java/awt/PopupMenu/PopupMenuLocation.java is unstable in MacOS machines

manukumarvs opened the pull request:
java/awt/PopupMenu/PopupMenuLocation.java seems to be unstable in MacOS machines, especially in MacOSX 12 & 13 machines. It seems to be a testbug as adding some stability improvements fixes the issue. It intermittently fails in CI causing some noise. This test was already problem listed in windows due to an umbrella bug JDK-8238720. This fix doesn't cover the Windows issue, so the problem listing in windows will remain same.

Fix:
Some stability improvements have been done and the test has been run 100 times per platform in mach5 and got full PASS.
Also updated the screen capture code, made frame undecorated, added some prints for better debugging.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8288415](https://bugs.openjdk.org/browse/JDK-8288415): java/awt/PopupMenu/PopupMenuLocation.java is unstable in MacOS machines


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10655/head:pull/10655` \
`$ git checkout pull/10655`

Update a local copy of the PR: \
`$ git checkout pull/10655` \
`$ git pull https://git.openjdk.org/jdk pull/10655/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10655`

View PR using the GUI difftool: \
`$ git pr show -t 10655`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10655.diff">https://git.openjdk.org/jdk/pull/10655.diff</a>

</details>

manukumarvs commented:
> So, what's the difference between this and #9187 and #10606? And did one of them have git issues?

No, #9187 got closed automatically by bot after being inactive for 8 weeks. And difference between #9187 and #10606 are,
1. Improved the screen capture logic in #10606 to capture the corresponding screen where the frame is displayed.
2. Made the frame undecorated as I have seen in some linux machines that, sometimes the mouse click inadvertently goes to the frame header which in turn fails the test(as it will not invoke the popup menu).

manukumarvs commented:
Any volunteers please.

manukumarvs commented:
Any volunteers for a review?.

DamonGuy commented on test/jdk/java/awt/PopupMenu/PopupMenuLocation.java:
```diff
@@ -106,22 +123,36 @@ private void show(MouseEvent e) {
         robot.mouseRelease(InputEvent.BUTTON3_DOWN_MASK);
-        int x = pt.x + frame.getWidth() / 2;
-        int y = pt.y + 130;
+        robot.waitForIdle();
```
When testing locally on my macOS machine, this waitForIdle seems to be the culprit for why this test takes much longer per iteration. Maybe worth looking into why?

manukumarvs replied:
Replaced robot.waitForIdle with robot.delay(200). I will check the possibility of raising a new bug.

aivanov-jdk replied:
Since no one expected for `robot.waitForIdle` to make the test so slow here, I suggest submitting a bug for evaluation. You have a working test case for this. The frame doesn't need to move across the screen.

manukumarvs replied:
I raised a new bug for this issue JDK-8299937

DamonGuy commented on test/jdk/java/awt/PopupMenu/PopupMenuLocation.java:
```diff
@@ -48,42 +54,52 @@ public final class PopupMenuLocation {
-                while (point.x
-                        < bounds.x + bounds.width - insets.right - SIZE) {
+            while (point.y < currentScreenBounds.y + currentScreenBounds.height - insets.bottom - SIZE) {
+                while (point.x <
+                           currentScreenBounds.x + currentScreenBounds.width - insets.right - SIZE) {
```
This might not be the most elegant way to abide by the rule to create a new line for long lines. I believe the line limit is 80 characters, which this still violates. And since it looks like you're trying to fix that here, might as well do it correctly.

aivanov-jdk replied:
I agree, the first `while` condition doesn't fit 80-column limit anyway, so the second may violate it too, and I think the readability only improves. Especially if you use `ScreenBounds` instead of `currentScreenBounds`.

Please, correct the positions of the closing braces: all of them are moved four spaces too far from their level, including the one which closes `main` method implementation.

DamonGuy commented on test/jdk/java/awt/PopupMenu/PopupMenuLocation.java:
```diff
@@ -95,6 +111,7 @@ public void mouseReleased(MouseEvent e) {
 
                 private void show(MouseEvent e) {
                     if (e.isPopupTrigger()) {
+                        System.out.println("Going to show popup "+pm+" on "+frame);
```
There seems to be a lack of consistency with spacing throughout the test. Spacing out the "+" from the quotations and the vars makes it cleaner.

DamonGuy commented on test/jdk/java/awt/PopupMenu/PopupMenuLocation.java:
```diff
@@ -106,22 +123,36 @@ private void show(MouseEvent e) {
-        int x = pt.x + frame.getWidth() / 2;
-        int y = pt.y + 130;
+        robot.waitForIdle();
+         y = y+50;
```
Another instance of the spacing issue. There seems to be more in this test, so it's probably best to fix them all.

aivanov-jdk replied:
This line has wrong indentation. I suggest using a compact syntax:

```suggestion
         y += 50;
```
or
```suggestion
         y += OFFSET;
```
if you create the `OFFSET` constant as I suggested above.

DamonGuy commented on test/jdk/java/awt/PopupMenu/PopupMenuLocation.java:
```diff
@@ -106,22 +123,36 @@ private void show(MouseEvent e) {
+    private static void captureScreen() {
+        try {
+            ImageIO.write(robot.createScreenCapture(currentScreenBounds), "png",
+                          new File("screen1.png"));
```
Is screen1.png the best name for the image? May be misleading if there are multiple screens, especially since this test iterates multiple locations.

aivanov-jdk replied:
It is definitely not. Either keep track of the current screen or use a generic `screen.png` file name.

I would also suggest, wrapping all the parameters to `write`:
```java
            ImageIO.write(robot.createScreenCapture(screenBounds),
                          "png",
                          new File("screen1.png"));
```
This way the `"png"` doesn't get “lost”.

DamonGuy reviewed (commented):
I ran the updated test in CI and locally, and it seems to be OK. My issue is that when I run locally, the test takes noticeably longer to execute and complete. I tried investigating the issue by implementing your changes incrementally, and I believe it's due to the waitForIdle after button3 presses. Maybe it's better to place a set amount of delay in ms here instead? Additionally, there are multiple spacing issues in the updated lines, and the header needs to be updated to have "2022" alongside the "2017" test date.

aivanov-jdk commented on test/jdk/java/awt/PopupMenu/PopupMenuLocation.java:
```diff
@@ -106,22 +123,36 @@ private void show(MouseEvent e) {
+        } catch (Exception exception) {
+            exception.printStackTrace();
+        }
+        action = false;
```
```suggestion
```
It's unneeded here and rather confusing.

aivanov-jdk commented on test/jdk/java/awt/PopupMenu/PopupMenuLocation.java:
```diff
@@ -106,22 +123,36 @@ private void show(MouseEvent e) {
+        try {
+            ImageIO.write(robot.createScreenCapture(currentScreenBounds), "png",
+                          new File("screen1.png"));
+        } catch (Exception exception) {
+            exception.printStackTrace();
```
```suggestion
        } catch (Exception e) {
            e.printStackTrace();
```
Why not use the standard name for the exception variable? Spelling the entire word doesn't add any value.

aivanov-jdk commented:
I also suggest removing the second asterisk at line 43, so that jtreg tags don't result javadoc warnings.

If you also remove `throws Exception` clause from `openPopup`, the test will become warning-free.

manukumarvs commented:
> I also suggest removing the second asterisk at line 43, so that jtreg tags don't result javadoc warnings.
> 
> If you also remove `throws Exception` clause from `openPopup`, the test will become warning-free.

Fixed the review comments and updated the test.

manukumarvs commented:
Accidentally closed the PR, reopening it.

aivanov-jdk commented on test/jdk/java/awt/PopupMenu/PopupMenuLocation.java:
```diff
@@ -47,43 +50,59 @@ public final class PopupMenuLocation {
-                        < bounds.x + bounds.width - insets.right - SIZE) {
+            Point point = new Point(screenBounds.x + insets.left,
+                                    screenBounds.y + insets.top);
+            while (point.y <
+                   screenBounds.y + screenBounds.height - insets.bottom -
+                   SIZE) {
+                while (point.x <
+                       screenBounds.x + screenBounds.width - insets.right -
+                       SIZE) {
```
```suggestion
            final int yBound = screenBounds.y + screenBounds.height
                               - insets.bottom - SIZE;
            final int xBound = screenBounds.x + screenBounds.width
                               - insets.right - SIZE;
            while (point.y < yBound) {
                while (point.x < xBound) {
```

I think this looks clearer and more readable than the wrapped condition.

aivanov-jdk commented on test/jdk/java/awt/PopupMenu/PopupMenuLocation.java:
```diff
@@ -47,43 +50,59 @@ public final class PopupMenuLocation {
     }
 
-    private static void test(final Point tmp) throws Exception {
+    private static void test(final Point tmp) {
```
```suggestion
    private static void test(final Point loc) {
```
Let's give a better name to the parameter.

aivanov-jdk commented on test/jdk/java/awt/PopupMenu/PopupMenuLocation.java:
```diff
@@ -47,43 +50,59 @@ public final class PopupMenuLocation {
             frame.add(pm);
             frame.pack();
             frame.setSize(SIZE, SIZE);
             frame.setVisible(true);
             frame.setLocation(tmp.x, tmp.y);
```
```suggestion
            frame.setLocation(tmp);
```
A `Point` object can be used directly.

aivanov-jdk replied:
```suggestion
            frame.setLocation(tmp);
            frame.setVisible(true);
```
You can pass a `Point` object to `setLocation` directly.

Setting the location before making the frame visible avoids flickering where the frame appears at 0, 0 and then moves to another location.

DamonGuy reviewed (approved):
Looks good. My main concern was the really long delay from waitForIdle. Glad a bug was raised for that and all of the minor issues were addressed.

manukumarvs commented:
/integrate

aivanov-jdk commented:
/sponsor
